plugins/inputs/prometheus/kubernetes.go

```diff
@@ -126,6 +150,169 @@ func (p *Prometheus) watch(ctx context.Context, client *k8s.Client) error {
 	}
 }
 
+func (p *Prometheus) cAdvisor(ctx context.Context, client *k8s.Client) error {
+	p.Log.Infof("Using monitor pods version 2 to get pod list using cAdvisor.")
+
+	nodeIP := os.Getenv("NODE_IP")
+	if nodeIP == "" {
+		return errors.New("The environment variable NODE_IP is not set. Cannot get pod list for monitor_kubernetes_pods using version 2.")
+	}
+
+	// Parse label and field selectors - will be used to filter pods after cAdvisor call
+	labelSelector, err := labels.Parse(p.KubernetesLabelSelector)
+	if err != nil {
+		p.Log.Errorf("Error parsing the specified label selector(s): %s", err.Error())
+	}
+	fieldSelector, err := fields.ParseSelector(p.KubernetesFieldSelector)
+	if err != nil {
+		p.Log.Errorf("Error parsing the specified field selector(s): %s", err.Error())
+	}
+	if fieldSelectorIsSupported(fieldSelector) {
+		p.Log.Errorf("A specified field selector is not supported as a field selector for pods")
+	}
+	p.Log.Infof("Using the label selector: %v and field selector: %v", labelSelector, fieldSelector)
+
+	// Set InsecureSkipVerify for cAdvisor client since Node IP will not be a SAN for the CA cert
+	tlsConfig := client.Client.Transport.(*http.Transport).TLSClientConfig
+	tlsConfig.InsecureSkipVerify = true
+
+	// The request will be the same each time
+	podsUrl := fmt.Sprintf("https://%s:10250/pods", nodeIP)
+	req, err := http.NewRequest("GET", podsUrl, nil)
+	if err != nil {
+		return err
+	}
+	client.SetHeaders(req.Header)
+
+	// Update right away so code is not waiting the length of the cAdvisorCallInterval initially
+	err = updateCadvisorPodList(ctx, p, client, req, labelSelector, fieldSelector)
+	if err != nil {
+		return err
+	}
+
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		case <-time.After(cAdvisorCallInterval * time.Second):
+			err := updateCadvisorPodList(ctx, p, client, req, labelSelector, fieldSelector)
+			if err != nil {
+				return err
+			}
+		}
+	}
+}
+
+func updateCadvisorPodList(ctx context.Context, p *Prometheus, client *k8s.Client, req *http.Request,
+	labelSelector labels.Selector, fieldSelector fields.Selector) error {
+
+	resp, err := client.Client.Do(req)
+	if err != nil {
+		return err
+	}
+	defer resp.Body.Close()
+
+	responseBody, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return err
+	}
+	cadvisorPodsResponse := podResponse{}
+	json.Unmarshal([]byte(responseBody), &cadvisorPodsResponse)
+	pods := cadvisorPodsResponse.Items
+
+	// Updating pod list to be latest cadvisor response
+	p.lock.Lock()
```

Comment:
registerPod already locks, and this will cause double locks and panic. You'll need to unlock before the loop, which means kubernetesPods will need a default value of an empty map or something. May be better off removing the locks from registerPod.

Reply:
It might simplify the logic, but I see you've resolved the main issue so I'll leave that up to you.